[RNMobile] Introduce Context Constants to Native

SiobhyB opened the pull request:
`Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4099

## Description

The Gallery block was [recently refactored](https://github.com/WordPress/gutenberg/pull/25940), for both the web and the mobile (native) apps, in order to pull in image blocks via [the InnerBlocks API](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/nested-blocks-inner-blocks/). 

As part of this refactor for the mobile apps, the following native-specific [block contexts](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/) were introduced in order to style the gallery's inner image blocks:
* `imageCrop`, was introduced to support the cropping of images. 
* `fixedHeight`, was introduced in https://github.com/WordPress/gutenberg/pull/36013 to explicitly set a fixed height for the images.

Similarly, as described in https://github.com/WordPress/gutenberg/issues/33531, an `allowResize` context was introduced on the web-side to disable the image block's resize option for galleries. 

<hr />

**Problem:** Additional contexts for customizing child blocks are going to be necessary. 

On the native side, for example, the settings icon is included within the block, unlike on the web, where the block's settings are separated out into the sidebar. The ability to hide the settings icon is therefore likely to be a fairly common use case for parent blocks. There are also likely to be other valid use cases for customizing inner blocks in the future.

Adding additional contexts isn't an ideal or scaleable solution. This has also been discussed in https://github.com/WordPress/gutenberg/issues/33531, with some suggestions outlined there.

<hr />

**Proposal:** With this PR, the idea of introducing context constants, as initially [suggested here](https://github.com/WordPress/gutenberg/pull/25940#discussion_r671011374), is built upon. 

`usesContextConstants` is provided as an alternative to `usesContext` and `providesContextConstants ` as an alternative to `providesContext`.

The aim is to explicitly indicate the purpose of these contexts and that they aren't serialised, as would normally be the case with contexts. It is also hoped that the introduction of context constants will make it easier to add these types of contexts in the future.

<hr />

**_Note: This PR is native-specific but a separate web-specific PR can be made, building on the work done here, if approved._**

## How has this been tested?

_Prequisite: Enable the flag for the refactored Gallery block. on mobile. At the time of writing, the way to enable the block on the Gutenberg Mobile demo app is to comment out [this section of code](https://github.com/WordPress/gutenberg/blob/5c9e84d66419c75900b8e6c5addc6eb2326e203e/packages/block-library/src/gallery/edit-wrapper.js#L33-L44)._ 

With the refactored Gallery block enabled, we should ensure that the changes introduced with this PR don't have any adverse effects:

* Navigate to the editor within either the Android or iOS app and add the Gallery block.
* Add a selection of images to the block. Verify that the images appear, which will in turn verify that the new `fixedHeight` context constant is working and hasn't caused any regressions (no images would be visible if the context wasn't working).
* Select the cog/gear icon to access the block's settings. Toggle the "crop images" tab on and off to verify that the images within the gallery change from being cropped, to not. This will confirm that the `imageCrop` context is still working as expected.

## Screenshots <!-- if applicable -->

The following screencast shows the images going from cropped to uncropped, as expected, to show the `imageCrop` context is working as intended following the changes in this PR:

https://user-images.githubusercontent.com/2998162/136827580-28d8e53a-4f90-42fc-bd37-ccb60c6c8514.mov

## Types of changes

This PR includes changes to code and the overall approach that's taken to adding contexts when needed to change inner blocks on the native-side, with the following high-level overview:

* `providesContextConstants` and `usesContextConstants` are added to block metadata upon registration [via the `blocks/src/api/registration.js` file](https://github.com/WordPress/gutenberg/pull/35510/files#diff-30ce126b90014d659302087765876898f16562219f796796ed407fdb346c8e4d). 
* `providesContextConstants` was [added to the object generated](https://github.com/WordPress/gutenberg/pull/35510/files#diff-f7f5a30681413bf9eb7368f36c36184b454dd2094c49c4c21930d414884d8624R35) by the `getBlockContext` function. There are also now checks in place [here](https://github.com/WordPress/gutenberg/pull/35510/files#diff-22dc9560179184651a38572ecad3bef26c968be9677771ddacf13fc8c131795bR34) and [here](https://github.com/WordPress/gutenberg/pull/35510/files#diff-b2f4aa1258856994e49ed89c4068795b171a9b7b7b90cc1440c272517b1792ecR140) to account for the existence of context constants on the native side.
* The Gallery and Image blocks have been refactored to use `fixedHeight` (an existing context, first implemented in https://github.com/WordPress/gutenberg/pull/36013) as an example of a context constant.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Enable native-specific metadata with block registration

With this commit, the 'packages/blocks/src/api/registration.js'  file is duplicated into a 'native' version. This is necessary as we'll be making native-specific changes to the metadata that's registered alongside a block.

SiobhyB committed:
Add "context constants" to block's metadata

'providesNativeInnerBlockContext' and 'usesNativeInnerBlockContext' are intended to replace the default 'providesContext' and 'usesContext'. These will be used to indicate times where a context is only used with the apps and constant (usually in use with inner blocks).

SiobhyB committed:
Enable use of 'providesNativeInnerBlockContext'

Following this commit, native blocks will now be able to use 'providesNativeInnerBlockContext' in order to provide a context to an inner block.

SiobhyB committed:
Enable use of 'usesNativeInnerBlockContext'

Following this commit, native blocks will now be able to use 'usesNativeInnerBlockContext' in order to use a context provided by a parent block.

SiobhyB committed:
Flag gallery block's native-specific contexts

As the 'imageCrop' context is only used with the app, it's flagged with 'providesNativeInnerBlockContext' as part of this commit.

SiobhyB committed:
Flag native-specific contexts used by image block

Following https://github.com/WordPress/gutenberg/pull/35510/commits/65c3dd696cf4c694e33b7f32e4a0154b4967de02, the 'imageCrop' context was flagged to indicate it's only used with the app. This commit builds on that work to reference it in the image block via 'usesNativeInnerBlockContext'.

SiobhyB committed:
Add logic to prevent error when looking up block context

We can only look up a block's context if 'blockType' is true. This commit therefore adds a check to ensure 'blockType' is true before creating a context object.

SiobhyB committed:
Add logic to prevent error when looking up block context

As with https://github.com/WordPress/gutenberg/pull/35510/commits/5a7d4143f4e74e9e7889267bce27349500d0a01e, we can only look up a block's context if 'blockType' is true. This commit addresses this issue in a second file.

SiobhyB commented:
👋  @geriux, I've requested your review as this includes changes to the Gallery block's contexts, which I know you worked on and might have a sense of whether this approach seems sound. Let me know if there's anyone else who you think should review, too.

@glendaviesnz, I also wanted to get your thoughts around whether this seems similar to what you had in mind when suggesting context constants.

As well as gathering your general thoughts on the code changes here, I'm keen to hear whether the naming of `usesNativeInnerBlockContext` and `providesNativeInnerBlockContext` makes sense to you. I also pondered on the following, and am very open to changing if you have ideas/feelings around anything that would make their purpose as clear as possible:

* `usesNativeContext` / `providesNativeContext`
* `usesNativeContextConstants` / `providesNativeContextConstants`

Thanks in advance! And please do let me know if anything needs clarifying around these proposed changes. 🙇‍♀️

glendaviesnz commented:
Thanks for the ping. A couple of thoughts after taking a look at this:

1/ I wonder if we need the native specific `providesNativeInnerBlockContext` namespace, and also further limiting it by tying it to innerBlocks. It seems to add complexity, and I think there is likely to be a need for other desktop parent blocks to handle this type of conflict as well - in which case we may end up having to duplicate context values between native and desktop namespaces. Maybe it is a case of tightening up the naming of the context values to avoid confusion, eg. `imageCrop` is meaningful in the context of the gallery block (although even there I don't think it is 100% clear what it means), but lacks meaning when passed down the context tree. Something like `cropToEqualSizeGrid` makes more sense at the gallery level, and it would make it clearer if the context that was passed down was something like `parentIsEqualSizeGrid` (still not perfect wording, just an example off the top of my head.

This would require a slightly different approach to context naming, which I haven't seen anywhere but may already be already be seen as a valid approach, and that is to use a different name for the context (name on left below) to the attribute (name on right), eg. 
```	
"providesContext": {
    "parentIsEqualSizeGrid": "cropToEqualSizeGrid"
}
```
To me this makes it easier to create names that are more meaningful when passed down than trying to come up with a single name that is accurate/meaningful in the parent and the children.

2/ The above naming discussion is separate to the idea of [context constants](https://github.com/WordPress/gutenberg/pull/25940#discussion_r671011374). That discussion related to context values that were not being serialized to the block attributes, and `imageCrop` is currently serialized to the gallery block attributes, so it is very standard to treat this as a block context value. 

The `allowResize` value in the gallery block however is not standard as it only ever uses the default value so is at no point serialized to the block attributes as context values were initially intended to be, hence the suggestion to be able to set a constant instead of adding an attribute that is never changed. 

Sorry, that is probably a little long winded. I may be overlooking some native specific issues as I haven't been involved directly in any of the native dev, but it seems like looking at how we name the existing context values rather than creating a new namespace may be an alternative way forward. The addition of a contextConstants namespace for the likes of `allowResize` could be handled separately to that.

SiobhyB commented:
@glendaviesnz, I really appreciate your thoughtfulness here, not long-winded at all!

It seems there's a good case for keeping `imageCrop` as a "regular" context. A problem is, though, that we'll still need to add additional contexts on the native side. These are a few that are currently in mind for the image block to be able to consume from parent blocks:

* `fixedHeight`: this will be provided by the core Gallery block and possibly other parent blocks in the future.
* `showSettings`: this will be provided by parent blocks to indicate whether the image's cog/settings icon should be displayed. This is fairly specific to native as the settings icon is included within the block, unlike on the web, where the block's settings are separated out into the sidebar.

The above contexts would be set to `true` by default, in a similar way to `allowResize`. I didn't include them in this PR in an attempt to keep its scope as narrow as possible, but it's definitely relevant as the aim is for the image block to be able to consume more of these types of contexts.

To move forward, and with your feedback in mind, do the following seem reasonable as next steps for me to take?

* Move `imageCrop` back to being a "regular" context and create a separate PR with the suggested tweaks to the naming.
* Rename `providesNativeInnerBlockContext`/`useNativeInnerBlockContext` to `providesContextConstants`/`usesContextConstants` and add `fixedHeight` or `showSettings` as examples for now.

glendaviesnz commented:
> * Move `imageCrop` back to being a "regular" context and create a separate PR with the suggested tweaks to the naming.

Over to you if you want to worry about a new PR for the name change at this point, was using that as an example to highlight issues with with some of the current naming rather than indicating that it was a critical change we should make. I imagine it will take some wider discussion if we do go that way, as I think there might be some that would prefer to keep the attribute and context names the same for consistency ... but if you are open to starting a conversation on that then an example PR is probably a good place to start. 

> * Rename `providesNativeInnerBlockContext`/`useNativeInnerBlockContext` to `providesContextConstants`/`usesContextConstants` and add `fixedHeight` or `showSettings` as examples for now.

Yeh, sounds good, will be interesting to see how this looks/works with a real example.

illusaen commented:
So I'm not as familiar with contexts as @glendaviesnz is, but one thought that I personally had was that there was a lot of repeat of 
```
const blockUsesContext = blockType && [
		...blockType.usesContext,
		...blockType.usesNativeInnerBlockContext,
	];
```
which could be extracted into a function that's used throughout these files.

SiobhyB committed:
Introduce 'fixedHeight' context constant

Following the discussion here, a new 'fixedHeight' context is added to the Gallery block with this commit: https://github.com/WordPress/gutenberg/pull/35510#issuecomment-942248340

This context will be provided by the core Gallery block and possibly other parent blocks in the future. Its default will always be true. It's currently used to set a fixed height on the native Gallery block.

SiobhyB committed:
Rename new context types to clarify purpose

The new context types are renmaed from "*NativeInnerBlockContext" to "*ContextConstants" following feedback here: https://github.com/WordPress/gutenberg/pull/35510#issuecomment-941869364

SiobhyB committed:
Keep 'imageCrop' as a 'regular' context

Discussion here: https://github.com/WordPress/gutenberg/pull/35510#issuecomment-941869364

SiobhyB committed:
Merge branch 'trunk' into rnmobile/try/creating-context-constants-for-gallery-block

SiobhyB committed:
Remove native-specific limitation

It should be possible for all platforms to make use of context constants. As such, this commit removes the native-specific registration file, so the PR's changes are now visible in the main registration file and apply to all platforms.

guarani commented on packages/block-library/src/image/block.json:
```diff
@@ -3,7 +3,8 @@
 	"title": "Image",
 	"category": "media",
-	"usesContext": [ "allowResize", "imageCrop" ],
+	"usesContext": [ "allowResize", "imageCrop"],
```
nitpick: the change on this line seems unintentional

SiobhyB replied:
Addressed in https://github.com/WordPress/gutenberg/pull/35510/commits/21d003f1651bb9ae2748595fdc78e9f247e0c838. Thanks!

guarani commented on packages/block-library/src/image/block.json:
```diff
@@ -3,7 +3,8 @@
 	"category": "media",
-	"usesContext": [ "allowResize", "imageCrop" ],
+	"usesContext": [ "allowResize", "imageCrop"],
+	"usesContextConstants": ["fixedHeight"],
```
nitpick: it seems this is missing the correct formatting (not sure if the VSC prettier plugin can auto-format)

SiobhyB replied:
Ah, yes, thank you for pointing that out! I have the prettier plugin but it doesn't seem to format the JSON files (maybe an issue with my settings, I'll look into it). I've addressed the issue in https://github.com/WordPress/gutenberg/pull/35510/commits/21d003f1651bb9ae2748595fdc78e9f247e0c838.

glendaviesnz commented on packages/block-library/src/gallery/block.json:
```diff
@@ -92,12 +92,19 @@
+		},
+		"fixedHeight": {
+			"type": "boolean",
+			"default": true
```
I think the idea with the constants was that they wouldn't be listed as block attributes because of the fact that they would be values that don't change and so aren't serialised in the block content. So if this is the case with fixed height this attribute wouldn't be here and below the constant would be defined as:
```
	"providesContextConstants": {
		"fixedHeight": true
	},
```

@talldan is this what you were imagining when you mentioned [this idea here](https://github.com/WordPress/gutenberg/pull/25940#discussion_r671011374)?

SiobhyB replied:
@glendaviesnz, thank you for this, what you described makes perfect sense and I've refactored the code accordingly. It'd be great to hear if you (and @talldan) think this is on the right path or if there's anything else that needs to be changed here. 🙇‍♀️

SiobhyB replied:
~_(Disclaimer: I'm aware of the error with the unit tests, which is directly related to the introduction of `providesContextConstants`. I plan to look into it after first gathering thoughts on the existing approach.)_~ Edit: I've updated the tests to resolve the error in https://github.com/WordPress/gutenberg/pull/35510/commits/8719c9e8cce8d940d3a180354722e63c7452fab6.

SiobhyB committed:
Fix incorrect formatting in JSON file

SiobhyB committed:
Refactor context constant, as serialization isn't needed

SiobhyB committed:
Refactor to prevent serialisation of context constants

As per the discussion here, part of the purpose of using context constant is to negate the need for serialisation, as the values won't change: https://github.com/WordPress/gutenberg/pull/35510#discussion_r732352036

As such, with this commit, the function that was evaluating the value of the context constants based on a block's attributes is refactored.

With this change, 'providesContextConstants' are added to the block's context object in its entirety, with no need to grab anything from the block's attributes.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/try/creating-context-constants-for-gallery-block

SiobhyB committed:
Update unit tests to account for context constants

glendaviesnz commented:
@SiobhyB  - I am busy with some 5.9 release things, but should hopefully have time to look at this before the end of the week.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/try/creating-context-constants-for-gallery-block

SiobhyB commented:
@glendaviesnz, just checking in, will you have time to look over this this week?

glendaviesnz commented:
> @glendaviesnz, just checking in, will you have time to look over this this week?

Thanks for the reminder @SiobhyB. I had a quick look and for some reason on web the `providesContextConstants` doesn't get set on the blockType - just appears as empty object for the Gallery even though it is set in the block.json - I haven't managed to work out why, will take another look tomorrow - but does it work on the web side for you?

Also, I wonder if we want a separate `useContextConstants` or if at the child level it just becomes a context value? I guess one advantage of separating it out is that you then know that these are values that are not going to change so no need to run through a useEffect, etc. to deal with changes.

@talldan when you suggested [this idea of contextConstants](https://github.com/WordPress/gutenberg/pull/25940#discussion_r671011374) where you thinking that they  would be defined separately at parent and child level or just merged in with the other usesContext values at the child level?

SiobhyB commented:
Thanks for this @glendaviesnz!

> I had a quick look and for some reason on web the providesContextConstants doesn't get set on the blockType - just appears as empty object for the Gallery even though it is set in the block.json - I haven't managed to work out why, will take another look tomorrow - but does it work on the web side for you?

I confess some ignorance on the web side of things, and had focused my testing efforts on mobile, so thank you for bringing this up! I'm seeing the same as you on the web, and haven't been able to figure out why just yet. I'll keep looking into it and report back any findings.

> Also, I wonder if we want a separate useContextConstants or if at the child level it just becomes a context value? I guess one advantage of separating it out is that you then know that these are values that are not going to change so no need to run through a useEffect, etc. to deal with changes.

Do you have a preference here? I can see benefits to both approaches and don't have a strong preference myself. Interested to hear @talldan's thoughts too.

glendaviesnz commented:
> Do you have a preference here?

I don't have a strong preference, let's wait and see what @talldan has to say. I will hopefully have more time next week to take another look at why it isn't working on web side.

SiobhyB commented:
From the mobile-side, we were able to use block supports to achieve what we needed to in https://github.com/WordPress/gutenberg/pull/36618 and https://github.com/WordPress/gutenberg/pull/36371. 

I'm going to go ahead to close this PR as I won't be working on it going forward and, following the discussion at https://github.com/WordPress/gutenberg/discussions/36582, it's unclear whether there's any other current need for context constants.
